compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELinalgFunctionOutlining.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/AMDAIEUtils.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "mlir/Conversion/FuncToLLVM/ConvertFuncToLLVM.h"
+#include "mlir/Conversion/FuncToLLVM/ConvertFuncToLLVMPass.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Hoisting.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+
+#define DEBUG_TYPE "iree-amdaie-linalg-function-outlining"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+unsigned uniqueOutlinedMatmul = 0;
```

Comment:
Global values like this arent a good idea. This can cause race conditions when run in multi-threaded context.

I think LLVM already has a way to deduplicate symbols. I dont know how to call it though.

Reply:
Hi @MaheshRavishankar @newling - so I tried looking into upstream and found a pass called `duplicate-function-elimination` whose description can be found [here](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Func/Transforms/Passes.td#L43-L52).

I agree we shouldn't use global variables like this - I have made the changes that incorporates the same logic without having to use a global variable and without using the above pass `duplicate-function-elimination`.

I did try to see how to use the above de-duplication pass and I'm attaching the patch in [this gist](https://gist.github.com/Abhishek-Varma/7bdad538e0080cde16442da6c133c0a4).

The main reason why I implemented the logic (for creating unique outlined function) the way I did with counter variables for corresponding compute op was :-
1. We aren't creating function definitions for every occurrence - and can reuse the existing one from the get-go - time complexity-wise, I believe what's being done in this patch is better.
2. Every unique instance from the same category of compute ops would have an incremental number associated with it which we can just inspect from the IR and figure the total unique instances too just by looking at the count.

Now, by using the above patch (`deduplicate-function-elimination`), both of the aforementioned points doesn't hold true. Here's why :-
Reg 1. Clearly we're creating N instances of functions for all N instances of LinalgOps - time complexity wise this is not good.
Reg 2. Now, the pass `duplicate-function-elimination` in order to perform deduplication, requires that the symbol name should be different even if the body is same. And the pass picks up the first instance and uses that to replace the other instances. What therefore happens now is :-
```
INPUT:
Assume we have five outlined functions for elemwise linalg.generic.
Suppose :-
linalg generic 1, linalg generic 2 and linalg generic 3 are equivalent.
linalg generic 4 and linalg generic 5 are equivalent.

OUTPUT:
func.func @generic_elemwise_0_outlined
func.func @generic_elemwise_3_outlined
```

Whereas with the logic I implemented we'd get :-
```
OUTPUT:
func.func @generic_elemwise_0_outlined
func.func @generic_elemwise_1_outlined
```

Apart from being able to reuse the function definition from the Symbol Table when we can, I'm being particular about this unique instance numbering too because I feel it helps (as it's incremental) and doesn't appear as a random count/number in the function name in the IR.